### PR TITLE
Iterable validation and flatmap

### PR DESCRIPTION
**Pull request by EivindAntonsen**
Two suggested extension functions.

### CODE REVIEW

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt [hunk 1]
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
...
 }
+
+/**
+ * Transforms an Iterable into nullable elements while only keeping the non-null elements.
+ */
+fun <T, R> Iterable<T>.flatMapIndexedNotNull(function: (index: Int, T) -> Iterable<R>?): List<R> {
+    return flatMapIndexedTo(ArrayList(), function)
+}
+
+private inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(destination: C,
```
Comment (LukasForst) [inline src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:508]: I'd suggest making this function public as the other `mapTo` functions are public - thus allowing inlining of `flatMapIndexedNotNull`

Comment (LukasForst) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:508]: keep in mind, that all public functions must be documented

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt [hunk 2]
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
...
+                                                                                    transform: (index: Int, T) -> Iterable<R>?): C {
+
```
Comment (LukasForst) [inline src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:510]: please format the file properly using IntelliJ (I intent to add some formatting/linting tool like detekt, but I haven't had time yet)
```
inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(
    destination: C,
    transform: (index: Int, T) -> Iterable<R>?
): C {
    forEachIndexed { index, element ->
        transform(index, element)?.let { elements ->
            destination.addAll(elements)
        }
    }
    
    return destination
}
```

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt [hunk 3]
```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
...
+/**
+ * Verifies the relationship between every element of an Iterable.
+ *
+ * Iterates through the elements invoking the validationFunction on each one.
+ * Returns false on the first element that does not pass the validation function, otherwise true.
+ *
+ * @param validationFunction is the accumulator function that verifies the elements.
+ */
+fun <S, T : S> Iterable<T>.validateElements(validationFunction: (acc: S, T) -> Boolean): Boolean {
```
Comment (LukasForst) [inline src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:528]: I'm not sure about name `validateElements` when this function provide possibility to validate relationship between ordered elements - so you're in a fact verifying that `n` and possibly `n+1` have some relationship

to me `validateElements` sounds like it provides a way how to validate collection of elements without any additional relationships

File: src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt [hunk 4]
```diff
@@ -224,4 +224,24 @@ internal class IterableExtensionsTest {
...
+        val incrementingNumbers = listOf(2, 3, 4, 5, 6)
+        val differenceIsMaxOne = incrementingNumbers.validateElements { previousNumber, nextNumber ->
+            nextNumber - previousNumber == 1
+        }
+
+        assertEquals(true, differenceIsMaxOne)
+    }
+
+    @Test
+    fun flatMapIndexedNotNull() {
```
Comment (LukasForst) [inline src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:239]: could you please add a test which utilizes filtering `null`? thanks!

File: src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt [hunk 5]
```diff
@@ -224,4 +224,24 @@ internal class IterableExtensionsTest {
...
     }
+
+    @Test
+    fun validateElements() {
+        val incrementingNumbers = listOf(2, 3, 4, 5, 6)
+        val differenceIsMaxOne = incrementingNumbers.validateElements { previousNumber, nextNumber ->
+            nextNumber - previousNumber == 1
+        }
+
+        assertEquals(true, differenceIsMaxOne)
```
Comment (LukasForst) [inline src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:235]: could you please add a test where the result is `false`? thanks!

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 2 above)
Comment (EivindAntonsen) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:510]: forgot to add code style schema for this fork, will correct it

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 3 above)
Comment (EivindAntonsen) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:528]: Agreed. Will find something more fitting.

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 2 above)
Comment (LukasForst) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:510]: > understandings of properly

heh, true, `properly` is not correct wording 😄  -- I'm using the formatting, IntelliJ is using, so I just let IntelliJ format everything before I commit it

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 3 above)
Comment (EivindAntonsen) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:528]: I think `foldValidated` is the closest match I can think of. It's essentially a fold with early return, no?

Comment (LukasForst) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:528]: sounds gut to me

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 2 above)
Comment (EivindAntonsen) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:525]: I hit format on both of those files, and it seems to have resulted in a lot of changes.

File: src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt (hunk 4 above)
Comment (EivindAntonsen) [reply src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:257]: Hi, in the recent commit I did include something that checked for nulls in the end result, but I'm not sure that is what you meant. If it wasn't, can you elaborate?

Comment (LukasForst) [reply src/test/kotlin/pw/forst/tools/katlib/IterableExtensionsTest.kt:257]: yup, that one, thanks!

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 2 above)
Comment (LukasForst) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt]: this formatting looks good to me 👍

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt (hunk 1 above)
Comment (EivindAntonsen) [reply src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt]: Added documentation similar to the standard kotlin lib as the functions themselves are really pretty similar.

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt [hunk 6]
```diff
@@ -497,3 +509,57 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
     }
     return sb.toString().restrictLengthWithEllipsis(totalLength)
 }
+
+/**
+ * Returns a single list of all not null elements yielded from results of [transform]
+ * function being invoked on each element of original collection.
+ */
+fun <T, R> Iterable<T>.flatMapIndexedNotNull(transform: (index: Int, T) -> Iterable<R>?): List<R> {
+    return flatMapIndexedTo(ArrayList(), transform)
```
Comment (LukasForst) [inline src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:518]: as the `flatMapIndexedTo` is inlined, this one can be too, please inline it, thanks!

File: src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt [hunk 7]
```diff
@@ -497,3 +509,57 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
...
+
+/**
+ * Validates the relationship between every element of an Iterable.
+ *
+ * Iterates through the elements invoking the validationFunction on each one.
+ * Returns false on the first element that does not pass the validation function, otherwise true.
+ *
+ * @param validationFunction is the accumulator function that verifies the elements.
+ */
+fun <S, T : S> Iterable<T>.foldValidated(validationFunction: (acc: S, T) -> Boolean): Boolean {
```
Comment (LukasForst) [inline src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt:545]: `inline` would be nice as well

**Review by LukasForst — CHANGES_REQUESTED** [review]
Hey, thanks for the update! Resolved everything, two more comments, about adding `inline` keyword.

Also, could you please rebase it on top of current master? There are some conflicts, thank you! Once conflicts are resolved, I'll approve the code.

**Comment by EivindAntonsen** [comment]
Hey I rebased before I added the inline (never done a rebase before, we've only used merge at my job), so there's one extra commit after. Does this mean I need another rebase?

**Comment by LukasForst** [comment]
> Hey I rebased before I added the inline (never done a rebase before, we've only used merge at my job), so there's one extra commit after. Does this mean I need another rebase?
thanks! now is just necessary to resolve conflicts with the master branch

[rebase](https://medium.com/datadriveninvestor/git-rebase-vs-merge-cc5199edd77c) is nice thing to do as it preserves linear history, however I don't enforce it, if you'd like to resolve the commits with merge, feel free

Right now, there's conflict in the `IterableExtensions` with current master branch -- there's an example how to resolve conflicts between forks [here](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036).

**Comment by EivindAntonsen** [comment]
Hm, my first time working on a fork here, not sure I did this right.

I merged this branch into master on my forked project, and then pushed. But it still does not look right here. Does that mean I should've made a pr from my forked master instead?

**Comment by LukasForst** [comment]
It's not necessary to make PR from your master branch, but it is necessary to rebase/merge the master branch of this repo to your `iterable-validation-and...` branch. 

Either by using [this](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks/38955036) approach or by updating your master directly using github (or at least I think it should be possible to update your master from my master using github)

**Comment by LukasForst** [comment]
You can create PR from my master to yours using this https://github.com/EivindAntonsen/katlib/compare/master...LukasForst:master link, once you have updated master, you merge it to your branch, which updates this PR and will be ready to be merged

**Comment by EivindAntonsen** [comment]
I think this should be it. :)
Let me know if something needs changing.

**Comment by LukasForst** [comment]
Oh, I looked into the logs and the Github action could not connect to gradle plugin repository, they have some problems, all green now! Great.

**Review by LukasForst — APPROVED** [review]
Approved, thank you!
